src/spec-lang/tc/typecheck.ts

```diff
@@ -1026,20 +865,17 @@ function tcIdVariable(expr: SId, ctx: STypingCtx, typeEnv: TypeEnv): SType | und
     return defNode.parameters[bindingIdx][1];
 }
 
-export function tcId(expr: SId, ctx: STypingCtx, typeEnv: TypeEnv): SType {
+export function tcId(expr: SId, ctx: STypingCtx, typeEnv: TypeEnv): TypeNode {
     if (expr.name === "this") {
         const contract = ctx[1] as ContractDefinition;
 
         expr.defSite = "this";
 
-        return new SPointer(
-            new SUserDefinedType(getUserDefinedTypeFQName(contract), contract),
-            DataLocation.Storage
-        );
+        return mkUserDefinedType(contract);
     }
 
     // First try to TC the id as a builtin type
-    let retT: SType | undefined = tcIdBuiltinType(expr);
+    let retT: TypeNode | undefined = tcIdBuiltinType(expr);
```

Comment:
Can the rest of this function, and the tcIdVariable() functions be signifficantly refactored using the new `resolveAny` function?

Reply:
Not quite. We can't replace `tcIdVariable` with `resolveAny` as `tcIdVariable` incorporates non-solidity definitions (let-bindings, user-defined functions and bindings in if_updated paths). We could replace the calls to `resolveByName` and `resolveTypeDef`, but we would need to propagate version and context into `tc()` which is a bunch of extra work. We should probably do this in the future though.